### PR TITLE
docker: support build args

### PR DESCRIPTION
**Pull request by psihachina**
This commit adds the support build args:

``` build {
        use "docker" {
                dockerfile="./simple_project/Dockerfile"
                build_args=["TEST_PATH=.", "BIN_PATH_2=./.bin/app"]
	}
    }
```

### CODE REVIEW

**Comment by psihachina** [comment]
Partially solves this problem #1228

**Comment by psihachina** [comment]
@krantzinator Hello. I would like to receive feedback. What do you think? Can I add something?

**Comment by AutomationD** [comment]
It would be great to get this merged. @briancain is there anything on the roadmap blocking this PR to get merged?

**Comment by briancain** [comment]
Nothing blocking that I know of @AutomationD 😄 I'll get this reviewed soon by myself or someone from the team @psihachina , thanks for opening the PR!

**Comment by briancain** [comment]
Hey @psihachina - I recommend fixing your branch and removing the `Merge branch` commit, what you had before was fine 👍🏻

**Comment by psihachina** [comment]
@briancain yes, did  stupid 😄, but I seem to have fixed it

File: builtin/docker/builder.go
Comment (briancain) [inline builtin/docker/builder.go:165]: It looks like we build up `buildArgs` here, but later only use `b.config.BuildArgs`? Maybe we could move this right above where `buildArgs` is used since we're only needing it conditionally later on after all the fallback if-statements.

**Review by briancain — CHANGES_REQUESTED** [review]
Hey there @psihachina ! Thanks for opening a pull request for this. Had 1 comment about moving something, and also I'd like to make sure `BuildArgs` is documented for the website. I couldn't push to your branch to update the PR because of permissions, so if you could, please add this to the `docker/builder.go` file:

```
commit 4c3fbdf0f79094805abb2a1d7dcea2b9dbfde13c (HEAD -> DOCKER-BUILD-ARGS)
Author: Brian Cain <bcain@hashicorp.com>
Date:   Mon May 3 13:11:27 2021 -0700

    Include docstring of buildargs for docker plugin on website

diff --git a/builtin/docker/builder.go b/builtin/docker/builder.go
index 6b34c051..c3e21062 100644
--- a/builtin/docker/builder.go
+++ b/builtin/docker/builder.go
@@ -121,6 +121,12 @@ build {
                ),
        )

+       doc.SetField(
+               "buildargs",
+               "An array of strings of build-time variables passed as build-arg to docker"+
+                       "or img for the build step.",
+       )
+
        return doc, nil
 }
```

Thank you! 😄

File: builtin/docker/builder.go
Comment (psihachina) [reply builtin/docker/builder.go:165]: @briancain, you mean it 

`

    } else {
		buildArgs := make(map[string]*string)
		if b.config.BuildArgs != nil {
			for _, arg := range b.config.BuildArgs {
				pair := strings.Split(arg, "=")
				buildArgs[pair[0]] = &pair[1]
			}
		}

		// No fallback, build with Docker
		step.Done()
		step = nil
		if err := b.buildWithDocker(
			ctx, ui, sg, cli, contextDir, relDockerfile, result.Name(), buildArgs,
		); err != nil {
			return nil, err
		}
	}
`

**Comment by denysvitali** [comment]
I would suggest using a `map[string] string` instead of a `[]string` for the build args 👍

File: builtin/docker/builder.go
Comment (denysvitali) [inline builtin/docker/builder.go:287]: I would strongly suggest quoting the parameters (so that `ENV_NAME="some value"` doesn't break the build) and to use an array of strings + join instead of a buffer, so that you can easily join them:
```go
strings.Join([]string{"--build-arg x=y", "--build-arg a=b"}, " ")
```

In your case:
```go
b := []string{}
for key, value := range m {
    b = append(b, fmt.Fprintf(b, "--build-arg %s=%s", key, value))
}
return strings.Join(b, " ")
```

**Review by denysvitali — CHANGES_REQUESTED** [review]
IMHO it still needs a little fix, but other than that LGTM

**Review by denysvitali — APPROVED** [review]
LGTM to me now! Thanks for the changes 🚀

File: builtin/docker/builder.go
Comment (briancain) [inline builtin/docker/builder.go:284]:
```suggestion
// Translates BuildArgs from a map of key vals into a string of '--build-arg key=vals'
func createBuildArgsString(m map[string]string) string {
```

We should have a more specific name for this func since this is what it's actually doing 😄

File: builtin/docker/builder.go
Comment (briancain) [inline builtin/docker/builder.go:290]:
```suggestion
	if m == nil {
		return ""
	}
	b := []string{}
	for key, value := range m {
		b = append(b, fmt.Sprintf("--build-arg \"%s=%s\"", key, value))
	}
	return strings.Join(b, " ")
}
```

I _think_ this join ends up adding extra whitespace to be passed to the `img` command if there are no buildargs defined. So maybe we can guard this on if BuildArgs are even defined.

File: builtin/docker/builder.go
Comment (briancain) [inline builtin/docker/builder.go:225]: I should of noticed this the first time I reviewed this PR, but why not make the config option `BuildArgs` a `map[string]*string` to begin with? I think this would save us a translation here. Sorry I didn't notice this the first time around!

File: .changelog/1346.txt
Comment (briancain) [inline .changelog/1346.txt:2]:
```suggestion
plugins/docker: Add support for build args for both docker and img builders
```

**Review by briancain — CHANGES_REQUESTED** [review]
Hey there @psihachina , thanks for all the updates! I left some feedback based on the latest round of updates. Let me know if you have any more questions! Otherwise I'd say this is good to go 😄

File: builtin/docker/builder.go
Comment (denysvitali) [reply builtin/docker/builder.go:290]: Are you sure?
https://play.golang.org/p/0BkcWnLhyKJ

Comment (briancain) [reply builtin/docker/builder.go]: Hrm, I had reproduced it yesterday in go playground to do this, which is why I had made the comment. Can't seem to reproduce this behavior this morning so looks like I was wrong about the extra whitespace 😄 Sorry about that @psihachina 😅

Comment (psihachina) [reply builtin/docker/builder.go]: @briancain bring it back?

Comment (briancain) [reply builtin/docker/builder.go]: Yeah you can delete the

```golang
if m == nil {
 return ""
}
```

Part. Sorry about the confusion! ❤️ 
Otherwise this PR looks good to go 😄

File: builtin/docker/builder.go
Comment (briancain) [inline builtin/docker/builder.go:281]:
```suggestion
```

Comment (briancain) [reply builtin/docker/builder.go:281]: Sorry about this! I was wrong yesterday on this feedback. Gonna add a quick suggested change and commit it for simplicity, otherwise this PR looks good to go ❤️ 👍🏻

**Review by briancain — APPROVED** [review]

**Comment by briancain** [comment]
Thank you @psihachina !! Looks great. Appreciate you taking the time to open this PR and thanks again for responding to all the PR feedback ❤️

**Pull request merged**
